File: peer_ios/peerapp/MainViewModel.swift
//
//  MainViewModel.swift
//  peerapp
//
//  Created by emblock on 12/17/25.
//

import Foundation
import Combine

// The same enum as in the Android project
enum ConnectionState {
    case disconnected
    case connecting
    case connected
}

class MainViewModel: ObservableObject {

    @Published var connectionState: ConnectionState = .disconnected
    @Published var peers: [String] = []

    private var peerClient: PeerClient?

    func connect(serverIp: String) {
        if peerClient == nil {
            connectionState = .connecting
            peerClient = PeerClient(
                serverIp: serverIp,
                onMessageReceived: { [weak self] message in
                    // The onMessageReceived callback is on a background thread,
                    // so we must dispatch UI updates to the main thread.
                    DispatchQueue.main.async {
                        // Assuming the message is a comma-separated list of peers
                        let newPeers = message.split(separator: ",").map(String.init)
                        self?.peers = newPeers
                        self?.connectionState = .connected
                    }
                }
            )
            peerClient?.start()
        }
    }

    func disconnect() {
        peerClient?.stop()
        peerClient = nil
        connectionState = .disconnected
        peers = []
    }
    
    // This is called automatically when the ViewModel is deallocated.
    deinit {
        peerClient?.stop()
    }
}

File: peer_aos/app/src/main/java/dev/bluelemonade/peerapp/MainViewModel.kt
package dev.bluelemonade.peerapp

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

enum class ConnectionState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED
}

class MainViewModel : ViewModel() {

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState

    private val _peers = MutableStateFlow<List<String>>(emptyList())
    val peers: StateFlow<List<String>> = _peers

    private var peerClient: PeerClient? = null

    fun connect(serverIp: String) {
        if (peerClient == null) {
            _connectionState.value = ConnectionState.CONNECTING
            peerClient = PeerClient(
                serverIp = serverIp,
                onMessageReceived = { message ->
                    // This is called from a background thread
                    viewModelScope.launch {
                        // A more robust solution would parse the message
                        // to differentiate between peer list updates and regular messages.
                        // For now, we'll assume any message is a peer update.
                        val newPeers = message.split(",")
                        _peers.value = newPeers
                        _connectionState.value = ConnectionState.CONNECTED
                    }
                }
            )
            peerClient?.start()
        }
    }

    fun disconnect() {
        peerClient?.stop()
        peerClient = null
        _connectionState.value = ConnectionState.DISCONNECTED
        _peers.value = emptyList()
    }

    override fun onCleared() {
        super.onCleared()
        peerClient?.stop()
    }
}
